packages/adapter/lib/adapter/adapter.js

```diff
@@ -8261,73 +8264,70 @@ function Adapter(options) {
             }
 
             if (this.autoSubscribe && typeof pattern === 'string') {
-                for (let s = 0; s < this.autoSubscribe.length; s++) {
-                    if (
-                        pattern === '*' ||
-                        pattern.substring(0, this.autoSubscribe[s].length + 1) === `${this.autoSubscribe[s]}.`
-                    ) {
+                for (const autoSub of this.autoSubscribe) {
+                    if (pattern === '*' || pattern.substring(0, autoSub.length + 1) === `${autoSub}.`) {
                         // remove this pattern from adapter list
-                        adapterStates.getState(`system.adapter.${this.autoSubscribe[s]}.subscribes`, (err, state) => {
-                            if (!state || !state.val) {
-                                return;
-                            }
-                            let subs;
-                            try {
-                                subs = JSON.parse(state.val);
-                            } catch {
-                                logger.error(
-                                    `${this.namespaceLog} Cannot parse subscribes for "${this.autoSubscribe[s]}.subscribes"`
-                                );
-                                return;
-                            }
-                            if (!subs[pattern]) {
-                                return;
-                            }
-                            if (subs[pattern][this.namespace] === undefined) {
-                                return;
-                            }
-                            subs[pattern][this.namespace]--;
-                            if (subs[pattern][this.namespace] <= 0) {
-                                delete subs[pattern][this.namespace];
-                            }
+                        let state;
+                        try {
+                            state = await adapterStates.getState(`system.adapter.${autoSub}.subscribes`);
+                        } catch {
+                            // ignore
+                        }
+                        if (!state || !state.val) {
+                            return;
+                        }
+                        let subs;
+                        try {
+                            subs = JSON.parse(state.val);
+                        } catch {
+                            logger.error(`${this.namespaceLog} Cannot parse subscribes for "${autoSub}.subscribes"`);
+                            return;
+                        }
+                        if (!subs[pattern]) {
+                            return;
+                        }
+                        if (subs[pattern][this.namespace] === undefined) {
```

Comment:
for security handle cases where value is no number at all? else "--" will also fail?
I do not trust "state values" that much because anyone can write them